In CI skip installing into local repository

nineinchnick opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->
I noticed the size of the cache package with the local Maven repo is about 7.5GB, near the 10GB limit. I think installing Trino packages into the local repository is not necessary, and might both save some time on copying and reduce the size of the cache. Downloading and extracting such a big cache takes about 1.5 minutes.

<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?
improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
ci

> How would you describe this change to a non-technical end user or system administrator?
n/a

## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

hashhar commented on .github/workflows/ci.yml:
```diff
@@ -60,7 +60,7 @@ jobs:
       - name: Maven Checks
         run: |
           export MAVEN_OPTS="${MAVEN_INSTALL_OPTS}"
-          $RETRY $MAVEN clean install -B --strict-checksums -V -T C1 -DskipTests -P ci -pl '!:trino-server-rpm'
+          $RETRY $MAVEN clean package -B --strict-checksums -V -T C1 -DskipTests -P ci -pl '!:trino-server-rpm'
```
`verify` is the stage just before `install`. `package` is even earlier.

Can we somehow get list of which plugins are bound to `install` phase to make sure nothing is skipped?

nineinchnick replied:
WDYM by bound to?

hashhar replied:
Plugins execute based on what lifecycle phase they are "bound" to. e.g. maven-compiler plugin is bound to "compile" phase, surefire is bound to "test" phase, failsafe plugin to "integration-test" phase.

I know I can look at pom.xml and see the `<phase>` under `<execution>` for a plugin but not sure if we can get a list from Maven which includes inherited stuff as well.

nineinchnick replied:
Oh, Maven plugins, all clear now. I'll replace `package` with `verify` everywhere, that should be much safer. I think I remember reading about some Maven commands to list all plugins and phases, I'll look it up.

nineinchnick replied:
Actually, I used `package` instead of `verify` everywhere we also add `-DskipTests`. But I'll check the plugins anyway.

nineinchnick replied:
I ran this `./mvnw fr.jcgay.maven.plugins:buildplan-maven-plugin:list-phase` and for `install` there's only `maven-install-plugin` for all modules. `verify` has `spotbugs-maven-plugin` and `maven-pmd-plugin`, so we need to run it instead of just `package`.

One example for `trino-hive`:
```
[INFO] Build Plan for trino-hive: 
validate -----------------------------------------------------------------------
    + maven-enforcer-plugin         | default               | enforce          
    + dependency-scope-maven-plugin | default               | check            
    + license-maven-plugin          | default               | check            
    + maven-checkstyle-plugin       | checkstyle            | check            
initialize ---------------------------------------------------------------------
    + git-commit-id-plugin          | default               | revision         
    + jacoco-maven-plugin           | default               | prepare-agent    
process-resources --------------------------------------------------------------
    + maven-resources-plugin        | default-resources     | resources        
compile ------------------------------------------------------------------------
    + maven-compiler-plugin         | default-compile       | compile          
process-test-resources ---------------------------------------------------------
    + maven-resources-plugin        | default-testResources | testResources    
test-compile -------------------------------------------------------------------
    + maven-compiler-plugin         | default-testCompile   | testCompile      
process-test-classes -----------------------------------------------------------
    + maven-dependency-plugin       | default               | analyze-only     
    + maven-dependency-plugin       | default               | analyze-duplicate
    + duplicate-finder-maven-plugin | default               | check            
    + modernizer-maven-plugin       | modernizer            | modernizer       
test ---------------------------------------------------------------------------
    + maven-surefire-plugin         | default-test          | test             
prepare-package ----------------------------------------------------------------
    + jacoco-maven-plugin           | report                | report           
package ------------------------------------------------------------------------
    + maven-jar-plugin              | default-jar           | jar              
    + maven-jar-plugin              | attach-tests          | test-jar         
    + maven-source-plugin           | attach-sources        | jar-no-fork      
    + maven-source-plugin           | attach-sources        | test-jar-no-fork 
verify -------------------------------------------------------------------------
    + spotbugs-maven-plugin         | default               | check            
    + maven-pmd-plugin              | default               | check            
install ------------------------------------------------------------------------
    + maven-install-plugin          | default-install       | install          
deploy -------------------------------------------------------------------------
    + maven-deploy-plugin           | default-deploy        | deploy        
```

nineinchnick commented:
@hashhar this is now ready for review. I changed install to package in a few places where we're only supposed to build the project and we run `mvn test` right after that. I changed it to `verify` in the `Maven Checks` job and I also added changes in the `error-prone-checks` job that @findepi suggested in #13257


I checked that with these changes the size of the local Maven repo is ~1GB and shouldn't grow much. Currently, it's ~7GB which is not sustainable.

findepi commented:
> I also added changes in the `error-prone-checks` job that @findepi suggested in #13257

Do you want me to merge or close that PR?

nineinchnick commented:
> Do you want me to merge or close that PR?

You can close it and merge this one.

findepi commented:
Thanks. just FYI, it's red now.

nineinchnick commented:
I made some wrong assumptions before and had to reevaluate my approach here. Where we run `mvn test` there has to be artifacts from other modules in the local repository, because the tested module has to be built. If I skip the build phase, it tries to run the artifact from the local repository, not the build directory.

I reverted replacing `install` with `package` in most jobs, except for `maven-checks`, which should be the only job where the cache can be saved. I added clearing the local repo in other jobs to ensure that.

nineinchnick commented:
@hashhar @findepi this is ready to go

findepi commented:
I don't know how our CI caching works. @hashhar can you PTAL?

hashhar commented:
How much time does our cache save us. If it's not a very large difference I'd probably opt to drop the cache instead.

hashhar commented on .github/workflows/ci.yml:
```diff
@@ -89,17 +89,17 @@ jobs:
           # Run Error Prone on one module with a retry to ensure all runtime dependencies are fetched
-          $RETRY $MAVEN ${MAVEN_TEST} -T C1 clean test-compile -P gib,errorprone-compiler -pl ':trino-spi'
+          $RETRY $MAVEN ${MAVEN_TEST} -T C1 clean verify -DskipTests -P gib,errorprone-compiler -pl ':trino-spi'
           # The main Error Prone run
-          $MAVEN ${MAVEN_TEST} -T C1 clean test-compile -P gib,errorprone-compiler \
+          $MAVEN ${MAVEN_TEST} -T C1 clean verify -DskipTests -P gib,errorprone-compiler \
```
This is the non-controversial part and we can merge it right away if extracted.

nineinchnick commented:
> How much time does our cache save us. If it's not a very large difference I'd probably opt to drop the cache instead.

Based on highly scientific tests, that is a single run of this branch

with cache:
![image](https://user-images.githubusercontent.com/795177/182850957-972c6183-5157-47d4-86d2-c6915ba80eb1.png)
without cache:
![image](https://user-images.githubusercontent.com/795177/182851058-a8709cb5-eb2b-47a2-a54a-3a66ce113017.png)

Note that this PR attempts to fix the cache to be under 1gb, instead of the 7gb we have now. Downloading that 7gb cache takes 2-5 minutes. I hope to get this under 30 seconds. Then it'll be beneficial, since it saves 50 seconds.

Note that not installing artifacts in a local maven repo should not be related to caching it or not. If we don't have to copy files around, we shouldn't.

hashhar commented:
I still see large cache being used - https://github.com/trinodb/trino/runs/7672132855?check_suite_focus=true#step:4:81

```
Cache Size: ~6914 MB (7249763541 B)
/usr/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/10163cbd-356c-43ed-98d8-a3350f756b3e/cache.tzst -P -C /home/runner/work/trino/trino
Cache restored successfully
```

nineinchnick commented:
I added a temporary commit with a change in the main pom to get a new cache key. When the run completes I'll trigger another one and that will demonstrate the changes.

nineinchnick commented:
I give up, I don't know what criteria Github uses when restoring the cache. It's supposed to check for a hash of all pom files, but I modified it and it still uses the cache from the previous workflow run:
```
Cache restored from key: setup-java-Linux-maven-0a53bc970763b2f9f964aba9ebd2fde72497ff9e8e884aac25d25409d362496b
```
where I computed the hash and its:
```
d14eb4b7c95b06aa930bc26818f8be71fb1aba18c6c758adbc7894f6c544a349
```
Maybe it's trying to be smart and getting some cache when there's no hit. Anyway, I used this action to verify the hash:
```
      - uses: nineinchnick/github-script@0a9984563e6199a44285d325a9921f1bff9517b2
        with:
          script: |
            const hash = await glob.hashFiles("**/pom.xml", {followSymbolicLinks: false}, true)
            core.info(hash)
```

findepi commented on .github/workflows/ci.yml:
```diff
@@ -164,6 +164,9 @@ jobs:
           retention-days: ${{ env.TEST_REPORT_RETENTION_DAYS }}
+      - name: Clean local Maven repo
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: rm -rf ~/.m2/repository
```
Add a code comment?

The commit changes some `install` into `package`, so worth noting why it's still worthwhile to delete `~/.m2/repository`.

Also, why ``~/.m2/repository`` and not eg `~/.m2/repository/io/trino` ?

nineinchnick replied:
This removes the whole local repo to avoid creating a cache entry from this job since it might not be the one with the most dependencies.

findepi replied:
Is it worth a code comment?
